examples/proposed/task-manager.w

```diff
@@ -0,0 +1,107 @@
+bring cloud;
+
+resource Utils {
+  extern "./helpers.js" static inflight uuid(): str;
+  extern "./helpers.js" static inflight sleep(ms: num): void;
+}
+
+/**
+ * A resource that allows you to run background tasks asynchronously,
+ * and check their status at any point in time.
+ *
+ * The function of the background task must be passed to the
+ * constructor so that its code and permissions can be deployed to
+ * the cloud when the TaskManager is created.
+ */
+resource TaskManager {
+  fn: cloud.Function;
+  state: cloud.Bucket;
+
+  init(handler: inflight (str): void) {
+    this.bucket = new cloud.Bucket();
+
+    let wrapper = inflight (id: str, input: str) => {
+      try {
+        let result = handler.invoke(message);
+        this.bucket.put("${id}/status", "done");
+        this.bucket.put("${id}/result", result);
+      } catch e {
+        this.bucket.put("${id}/status", "failed");
+        this.bucket.put("${id}/result", e);
+      }
+    };
+
+    this.fn = new cloud.Function(wrapper);
+  }
+
+  /**
+   * Starts a background task with the given input, and returns an id
+   * that can be used to check the status of the task.
+   */
+  inflight start(input: str) {
+    let id = Utils.uuid();
+    this.fn.invoke_async(id, input);
```

Comment:
Kinda unrelated, but @Chriscbr you are still the guy to ask

why do we have `fn.invoke_async()` vs `defer fn.invoke()`

Reply:
You can use `defer fn.invoke()` as long as you `await` it later. (If you do not call `await` on the variable after you `defer` it, I think Wing should give you a compile error).

However for many applications you might not want to `await` it in your main loop -- for example, suppose you want a "GET /hello" request to start a job on the backend but return 200 as long as the job successfully started. In that case, I think `fn.invoke_async()` is a better option. `fn.invoke_async()` just waits for the cloud provider that says "Success, I started the function invocation" so that if it fails due to scaling/throttling etc. you can retry it.

It's also nice in that first time Wing users can use `fn.invoke_async()` without needing to learn about `defer` syntax (less cognitive load).